### PR TITLE
[#6256,#7220,#7519,#7545,#7546,irods/irods_resource_plugin_s3#2156] do not explicitly build against libc++ (main)

### PR DESCRIPTION
**Pull request by SwooshyCueb**
Merge irods/externals#212 and irods/irods_development_environment#128, then update externals package repos before merging
In service of #6256
In service of #7220
In service of #7519
In service of #7545
In service of #7546
In service of irods/irods_resource_plugin_s3#2156

Companion PRs:
- irods/irods_rule_engine_plugin_audit_amqp#133
- irods/irods_api_plugin_genquery2#89
- irods/irods_capability_storage_tiering#260
- irods/irods_microservice_plugins_curl#63
- irods/irods_client_http_api#256
- irods/irods_netcdf#51
- irods/irods_resource_plugin_s3#2170
- irods/irods_rule_engine_plugin_logical_quotas#105
- irods/irods_rule_engine_plugin_python#192
- irods/irods_rule_engine_plugin_metadata_guard#73

### CODE REVIEW

**Comment by SwooshyCueb** [comment]
Marked ready for review

File: .github/workflows/build-irods-centos.yml [hunk 1]
```diff
@@ -14,27 +14,27 @@ jobs:
...
               yum -y update
               yum -y install \
-                irods-externals-avro1.11.0-2 \
-                irods-externals-boost1.81.0-0 \
+                irods-externals-avro1.11.0-3 \
+                irods-externals-boost1.81.0-1 \
                 irods-externals-catch22.13.8-0 \
-                irods-externals-clang-runtime13.0.0-0 \
-                irods-externals-clang13.0.0-0 \
+                irods-externals-clang-runtime13.0.1-0 \
```
[the remark this answers is not in the document]
Comment (SwooshyCueb) [reply .github/workflows/build-irods-centos.yml:28]: No, good catch

[the remark this answers is not in the document]
Comment (SwooshyCueb) [reply .github/workflows/build-irods-centos.yml]: Fixed

**Comment by SwooshyCueb** [comment]
> Would it be worth splitting out the spdlog changes from the "Do not explicitly build against libc++" commit? Feels like the libc++ thing should be separate given its significance. I do not feel strongly about this.

The libstdc++ commit is the one updating externals, which necessitates the changing of the spdlog target, so it makes sense to me to keep them in the same commit.

File: cmake/Modules/UseLibCXX.cmake [hunk 2]
```diff
@@ -14,15 +16,15 @@ cmake_minimum_required(VERSION 3.6.0 FATAL_ERROR)
...
-  set(IRODS_BUILD_AGAINST_LIBCXX_DEFAULT ON)
-else()
-  set(IRODS_BUILD_AGAINST_LIBCXX_DEFAULT OFF)
+string(TOUPPER "${PROJECT_NAME}" PROJECT_NAME_UPPER)
+if (NOT PROJECT_NAME_UPPER STREQUAL "IRODS")
+  message(DEPRECATION "The UseLibCXX CMake module is deprecated and will be removed in a future relase of iRODS.")
 endif()
 
-set(IRODS_BUILD_AGAINST_LIBCXX ${IRODS_BUILD_AGAINST_LIBCXX_DEFAULT} CACHE BOOL "Try to build against libc++ instead of libstdc++.")
+set(IRODS_BUILD_AGAINST_LIBCXX OFF CACHE BOOL "[DEPRECATED] Try to build against libc++ instead of libstdc++.")
```
[the remark this answers is not in the document]
Comment (SwooshyCueb) [reply cmake/Modules/UseLibCXX.cmake:24]: > Does this mean that we are defaulting to building using libstdc++?

Yes. All distros we support build against libstdc++ by default.

> Do we need to switch which externals we are using based on this flag (i.e. the libcxx variants)?

Good idea.

**Comment by SwooshyCueb** [comment]
Since I'm going to be changing our externals to depend on the value of `IRODS_BUILD_AGAINST_LIBCXX`, I've decided to also address #7545 and #7546 in this PR.
Marking as draft again for now.

**Comment by SwooshyCueb** [comment]
Marking as ready for review again. Note the two new commits.

**Review by alanking — APPROVED** [review]
New changes look good to me. Let's wait for another approval before we squash and all that.

**Review by korydraughn — COMMENTED** [review]
Looks good.

There are multiple comments containing TODO. Let's remove them and/or replace them with issues.

You can reference the issue number in the code if that's helpful.

**Comment by SwooshyCueb** [comment]
> There are multiple comments containing TODO. Let's remove them and/or replace them with issues.
>
> You can reference the issue number in the code if that's helpful.

I've added references to the issue number to the TODO comments.

File: CMakeLists.txt [hunk 3]
```diff
@@ -16,25 +16,34 @@ endif()
...
-endmacro()
-
+# This is duplicated from UseLibCXX.cmake, since we need it before UseLibCXX can be included.
+# TODO: Remove this when UseLibCXX.cmake is removed (#7547).
+set(IRODS_BUILD_AGAINST_LIBCXX OFF CACHE BOOL "[DEPRECATED] Try to build against libc++ instead of libstdc++.")
+
+# TODO: At present, our split externals packages have parity. Eventually, this will change, likely due to
+# needs of the iRODS project, or due to the retirement of the -libcxx packages. When this happens, or when
+# UseLibCXX.camke is removed (#7547) (whichever happens first), support for automatically switching exernals
+# packages based on IRODS_BUILD_AGAINST_LIBCXX should be removed.
```
Comment (korydraughn) [inline CMakeLists.txt:26]: Now that we have issue numbers in the comments, please remove all uses of the word TODO.

That is what I was trying to express.

Comment (SwooshyCueb) [reply CMakeLists.txt]: Done

**Review by korydraughn — APPROVED** [review]
Pound it.

**Comment by korydraughn** [comment]
Pound if ready.

**Comment by SwooshyCueb** [comment]
\#'d

**Pull request merged**